Add RAMSETE pathfollowing support

5t0n3 opened the pull request:
The new `FollowTrajectory` command can be used to make the robot follow a trajectory, and the `TrajectoryCommandGroup` builds on this by allowing for other commands to run along a trajectory, being triggered at specific points. 

PathPlannerLib was also installed as a dependency, since PathPlanner provides a nice UI for generating trajectories.

Commit:
Add FollowTrajectory command (untested)

Commit:
Add an example trajectory-following auto

PathPlannerLib was also installed to allow for the loading of its
generated trajectories.

Commit:
Trajectory following works* now!

The robot no longer spins out when it tries to follow a trajectory, but
we still have to test accuracy. The spinning out was due to us swapping
the left and right drive Spark Maxes, which was fixed in this commit.
This also means the right stick powers for turning no longer have to be
negated.

(post-rebase note: We just had the wrong kFF value on the spark maxes;
the velocities should have been in meters per second. As of this commit
it doesn't actually work, but this is fixed in the next few commits.)

Commit:
Make AlignWithHub extend TurnDegrees

They have extremely similar behaviors, so I figured that this could be
done without issue.

(post-rebase note: this didn't actually turn out to be true, but was
fixed in later commits.)

Commit:
Move kinematics from Constants to FollowTrajectory

I also documented the FollowTrajectory command since I didn't do that
before.

Commit:
Trajectory following is now more accurate

When using the Spark Max closed loop velocity control, even after
reducing feedforward the bot would consistently overshoot. Trajectory
following seems more accurate after transitioning to voltage-based
control using feedforward/PID values obtained from SysId.

Also added another test trajectory called Squiggle, which moves the
robot forward along a curve and then backward along a different curve.

Commit:
Transition to a hybrid drive motor control method

I was reading through the FRC server, and someone mentioned that they
used WPILib feedforward with Spark Max closed-loop velocity control on
their drive motors. I thought that would be interesting to try, since
it could alleviate the pain of tuning kFF on the Spark Maxes, as well as
allowing acceleration to be accounted for. This commmit changes both the
DriveStraightProfiled (now DriveMeters) and FollowTrajectory commands to
utilize this form of hybrid control. As of now it is untested, but I
will do that when I get a chance.

Commit:
Reset previous velocities after finishing movement

Before, running two DriveMeters/FollowTrajectory commands one after the
other could have resulted in incorrect acceleration readings. This is
something that I forgot to take into account in my previous commit.
Right now this is just exposed as another method on DriveSubsystem, but
I think that should be changed before merging this branch.

Commit:
Transition to state-space feedforward for driving

Using a development version of SysId, I was able to get reasonable
values for kV and kA (for angular motion), so I figured that I could
integrate them into driving.

I also moved subsystem resetting (stopping motors, resetting the drive
feedforward) to a method within RobotContainer to avoid having to repeat
those method calls everywhere.

I also have no idea if these changes work, but I'll definitely be
testing them on Monday :)

Commit:
Increased readability of DriveSubsystem.tankDriveVelocities

Also increased the maximum speed/acceleration to theoretically
possible values based on the drivetrain current draw. These still have
to be tested, however.

Commit:
Reduce max speed/acceleration & format w/ spotless

After messing around with the robot simulation, the previous maximum
velocity & acceleration seemed to lead to required voltages above 12V,
which obviously aren't possible on a real robot. I also figured that
we should leave some leeway for other mechanisms, since we don't only
have the 6 drivetrain motors on the bot.

Commit:
Bring back AlignWithHub from main since it got lost in the rebase

Commit:
Added TrajectoryCommandGroup

This should work as a command group that executes other commands as the
robot drives along a trajectory, although it still has to be tested.

Commit:
Commands are now properly initialized in TrajectoryCommandGroup

Commit:
Add an example TrajectoryCommandGroup autonomous

It should drive along a slightly curved trajectory and enable the intake
near the end, but it has yet to be tested.

Commit:
Fix an exception in TrajectoryCommandGroup

I didn't realize that modifying a list while iterating over it would
throw a ConcurrentModificationException, although it makes sense. I
modified the affected loop so it iterates over a ListIterator, which
does allow for the modification of a list as its iterated over.

Spotless formatting was also applied, since I haven't done that in a
bit.

Commit:
TrajectoryCommandGroup subsystem requirement logic now mirrors ParallelCommandGroup

Since there is no way to know when a command will end, I decided to
delegate those decisions to the user by recommending the use of
ProxyScheduleCommands, since those will interrupt each other
automatically when necessary.

Commit:
Make naming in TrajectoryCommandGroup more clear

Commit:
Fixed another ConcurrentModificationException

This time it was occuring while iterating the untriggered command map in
TrajectoryCommandGroup.execute().

I also realized that IntakeCargo never required the intake/indexer
subsystems, so I added that in since it's pretty necessary.

Commit:
Use RAMSETE for driving straight distances as well

Instead of modifying DriveMeters, I decided to move the command into
DrivetrainSubsystem via a factory method. I think going forward this
would be a good idea instead of subclassing things like
ProfiledPIDCommand, for example, since the main thing we do is just call
the super() constructor.

Commit:
Remove unused DrivetrainSubsystem methods

tankDriveVolts was superseded by tankDriveVelocities, and getWheelSpeeds
seemed to be unreliable with RAMSETE control due to encoder velocity
latency with the Spark Maxes.

The updateOdometry method now returns the robot's new pose, negating
the need for a call to getPose() in DrivetrainSubsystem's periodic
method.

Commit:
Return to a TrapezoidProfileCommand for driving straight

I didn't realize it earlier but attempting to generate a straight
trajectory that spans a short distance seems to cause malformed spline
errors for some reason. Because we do PID control on the Spark Maxes
themselves, I think that just using a TrapezoidProfileCommand for
driving distances should work fine (although this should be tested).

Commit:
Removed an unnecessary `this`

Commit:
Bring back static feedforward component

This seemed to increase the forward/backward distance accuracy a bit,
but at the cost of leading to left/right overshoots for Test Path.

I also changed Test Path to drive forward and backward along the same
path, which makes verifying accuracy easier. That way I only have to
measure a couple inches instead of the full path distance.

nolanhergert commented on src/main/java/frc/robot/subsystems/DrivetrainSubsystem.java:
```diff
@@ -34,6 +46,14 @@ public class DrivetrainSubsystem extends SubsystemBase {
+
+  // Feedforward based on above linear system
+  private final LinearPlantInversionFeedforward<N2, N2, N2> m_feedforward =
+      new LinearPlantInversionFeedforward<>(m_drivetrainPlant, 0.02);
```
What is the .02 here referring to? Might want to make it a constant in a header file instead for readability?

Looks like static offset, but it took me a few minutes.

5t0n3 replied:
That's the controller period, which in this case is just the default robot loop period (20ms). I can move that to the constants file to make that more clear :)

nolanhergert commented on src/main/java/frc/robot/commands/autonomous/TrajectoryCommandGroup.java:
```diff
@@ -0,0 +1,188 @@
+    // Drive along the trajectory
+    m_trajectoryCommand.execute();
+
+    // A ListIterator allows for the modification of a list while iterating over it
```
When is this occurring again? (modifying while iterating)

5t0n3 replied:
When a command finishes it's removed from the currently executing list (this actually happens on [line 86](https://github.com/AEMBOT/FRC_2022/pull/6/files/dfc8283251690b391a0969e2b1720484ed818ac1#diff-78c97e56a7335e440b23b0c7f24702eae436bce46ba902af90caf1a7eb98f1f9R86) of TrajectoryCommandGroup.java). I could also store the finished commands in a separate list and remove them after finishing iterating, but removing them during the iteration made more sense in my head.

nolanhergert commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -178,13 +168,28 @@ public static final class TurnPID {
+      public static final double kRamseteZeta = 0.7;
+
+      // PID/Feedforward constants obtained from SysId
+      // public static final double kPDriveVelocity = 0.059288;
```
I think you can remove these commented values

nolanhergert commented on src/main/java/frc/robot/commands/autonomous/TrajectoryCommandGroup.java:
```diff
@@ -0,0 +1,188 @@
+      Command command = poseCommand.getKey();
+
+      // Throw an exception if two commands share a subsystem dependency
+      if (!Collections.disjoint(m_requirements, command.getRequirements())) {
```
Say we need to turn off and turn on our shooter multiple times along the path, is this possible right now? (since both commands require the same subsystem)

5t0n3 replied:
That is possible if you wrap all of the commands in proxy schedule commands, since they don't inherit the dependencies of their inner commands. 

The reason I did it this way was because there isn't a way to programmatically verify when (or rather where) a command will end, so I figured I'd leave it to the programmer to reason through. Also, if two commands end up trying to run at the same time, the command scheduler will handle the command interruption logic, rather than that having to be handled within TrajectoryCommandGroup.

nolanhergert replied:
I see you have a nice exception message. Sorry!

nolanhergert commented on src/main/java/frc/robot/subsystems/DrivetrainSubsystem.java:
```diff
@@ -162,6 +185,39 @@ public void arcadeDrive(double speed, double rotation, boolean squareInputs) {
+    double rightVolts = feedforwards.get(1, 0) + Math.signum(rightVelocity) * kSVolts;
+
+    // Command the motors at the above feedforward voltages, using PID to correct for error
+    m_leftController.setReference(leftVelocity, CANSparkMax.ControlType.kVelocity, 0, leftVolts);
```
Can you explain this to me a little more? I don't get why we need to send both velocity and voltage down to the controller.

5t0n3 replied:
Yeah sure! The voltage is just the feedforward voltage supplied to the motor controller, which allows us to account for acceleration and static friction in the feedforward calculations (as well as use `LinearPlantInversionFeedforward`). The velocity is fed in for (velocity) PID control to be done on the Spark Maxes themselves, since as we've discussed they refresh at a higher rate than the main robot loop.

There is a way to get a higher refresh rate on the rio itself, but it takes a lot of boilerplate to get working in command-based.

Let me know if that all made sense :)

nolanhergert replied:
Yep!

nolanhergert reviewed (approved):
Congratulations on getting this working Zane! This represents a lot of work on your part, I'm sure :)

Commit:
Reorganize drivetrain constants

Right now, the only inner class it has is TurnPID, although I'll
probably work on removing that eventually.

I also fixed TurnDegrees (I think), since I changed its feedforward to
be in terms of volts without changing the rest of the logic. That
transition is outside of the scope of this commit.

Commit:
Rearrange example TrajectoryCommandGroup auto

The homing and lowering of the intake seemed inconsistent when it was
run as part of the TrajectoryCommandGroup. I realized that those things
should probably be run before the robot starts moving, so I changed the
declaration to reflect that.

I also removed a commented out section of AlignWithHub, since there was
no reason to keep it around.

As always, Spotless was applied for formatting.

nolanhergert commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -15,6 +15,9 @@
  */
 public final class Constants {
+  // The default period of the robot loop is 20 milliseconds (0.02 seconds)
+  public static final double kRobotLoopPeriod = 0.02;
```
Can get this from TimedRobot.java: https://github.com/wpilibsuite/allwpilib/blob/main/wpilibj/src/main/java/edu/wpi/first/wpilibj/TimedRobot.java#L67

Commit:
Set motors to brake mode again

Commit:
Slight constants changes
